### PR TITLE
optimize doc-level monitor execution workflow for datastreams

### PR DESCRIPTION
**Pull request by sbcd90**
*Issue #, if available:*
#1290 

*Description of changes:*
This pr optimizes the doc-level monitor execution workflow for data streams.

The optimization proposed only considers the `current write index, any new read-only indices created after last monitor execution run & the last write index of last monitor execution run` for current monitor execution run.
It ignores `read-only indices` that may have been created prior to the last monitor execution run.
 
The below diagram shows the difference.

**Earlier**
Screenshot 2023-11-10 at 6 00 48 PM

**Now**
Screenshot 2023-11-10 at 5 58 30 PM

*CheckList:*
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt [hunk 1]
```diff
@@ -207,11 +207,25 @@ class DocLevelMonitorQueries(private val client: Client, private val clusterServ
 
         // Run through each backing index and apply appropriate mappings to query index
         indices.forEach { indexName ->
-            val concreteIndices = IndexUtils.resolveAllIndices(
+            var concreteIndices = IndexUtils.resolveAllIndices(
                 listOf(indexName),
                 monitorCtx.clusterService!!,
                 monitorCtx.indexNameExpressionResolver!!
             )
+            if (IndexUtils.isAlias(indexName, monitorCtx.clusterService!!.state()) ||
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:215]: if I use any alias, there can be multiple writeIndices. why are we assuming there can be only one writeIndex. That applies to datastream only

plz add tests for alias which points to multiple write indices and modify handling as required. 

I would explore modifying monitor metadata as required to handle aliases with multiple write indices.

**Comment by eirsep** [comment]
plz make change to handle indices queried if the monitor metadata section is empty. Only write index and indices created after creation_time timestamp should be considered for querying .

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt (hunk 1 above)
Comment (sbcd90) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:215]: An `index/datastream alias` can have only `1 write index`. https://www.elastic.co/guide/en/elasticsearch/reference/current/aliases.html#write-index
`Common Problems` section - https://opster.com/guides/elasticsearch/glossary/elasticsearch-alias/
An `index pattern` can have multiple write indices though.

**Review by eirsep — COMMENTED** [review]
if they have 2 data streams a1* and a2* but customer creates index on a* index pattern, there will be multiple write indices for one input in `indices` section of monitor config. the changes in new PR might break the experience for this scenario
it’s ok if PR doesn’t optimize for that scenario but should not cause new bug

**Comment by sbcd90** [comment]
> if they have 2 data streams a1* and a2* but customer creates index on a* index pattern, there will be multiple write indices for one input in `indices` section of monitor config. the changes in new PR might break the experience for this scenario it’s ok if PR doesn’t optimize for that scenario but should not cause new bug

the optimization doesnt add a new bug. https://github.com/opensearch-project/alerting/blob/510e66acfb7fc30248097e77672a4a02f9081e61/alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt#L1398 
a follow-up pr need to be created for using this optimization in index patterns.

**Review by eirsep — APPROVED** [review]

**Review by praveensameneni — APPROVED** [review]

**Pull request merged**